### DISCUSSION TITLE
Comfusion about implementation of XPBD FEM constraint

### DISCUSSION THREAD

**Discussion by blacksino** [InteractiveComputerGraphics/PositionBasedDynamics#137] [Q&A] [answered]
Hi, I have been reading the code of the implementation of the XPBD FEM constraint, however, it seems to be a bit different from the formula from the XPBD paper.

    Real sum_normGradC =
            invMass0 * gradC[0].squaredNorm() +
            invMass1 * gradC[1].squaredNorm() +
            invMass2 * gradC[2].squaredNorm() +
            invMass3 * gradC[3].squaredNorm();

    C = sqrt(2.0 * C); // why?

    Real alpha = static_cast<Real>(1.0) / (youngsModulus * dt * dt);
    sum_normGradC += C * C * alpha; // why multiply alpha with C**2?

    if (sum_normGradC < eps)
        return false;

    // compute scaling factor
    //const Real s = C / sum_normGradC;
    const Real s = (C * C + C * alpha * multiplier) / sum_normGradC; // Why is there an extra C multiplied on numerator?

   //update mulitiplier,aka lamda?

    corr0 = -s * invMass0 * gradC[0];
    corr1 = -s * invMass1 * gradC[1];
    corr2 = -s * invMass2 * gradC[2];
    corr3 = -s * invMass3 * gradC[3];`

**Comment by janbender** [comment, marked as answer]
Hm, the names of the variables are not self-explaining due to copy & paste. I have to fix that. The explanation is that we use the potential energy: 

U = 0.5 C^T alpha^-1 C  (Eq. 5 in XPBD paper)

In our original paper we used C= U in the PBD framework. To adapt this to XPBD we now use: 

C = sqrt(2 U') 

with U = alpha^-1 U' and alpha = 1/youngsModulus. 
This yields: 

grad C = 1/C grad U'

This is where the additional C's come from. In the code grad U' has the name "gradC" which is quite misleading which comes from lazy copy & pasting stuff. 

I hope that clarifies this a bit.

**Reply by blacksino** [reply]
thank you, that helps a lot!
